### PR TITLE
FRR-BGP Extended Unified Management Framework

### PR DESCRIPTION
**Pull request by karthikarum**

### CODE REVIEW

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md [hunk 1]
```diff
@@ -0,0 +1,879 @@
...
+### 4.2.2 CLI Show Sequence - State/Statistics
+![FRR-BGP-CLI-SHOW-SEQUENCE2](images/frr-bgp-cli-show-sequence22.jpg)
+
+## 4.3 REST Get Sequence
+
+### 4.3.1 REST Get Sequence - Config information
+![FRR-BGP-REST-GET-SEQUENCE1](images/frr-bgp-rest-get-sequence1.jpg)
+
+### 4.3.2 REST Get Sequence - State/Statistics
+![FRR-BGP-REST-GET-SEQUENCE2](images/frr-bgp-rest-get-sequence2.jpg)
```
Comment (anand-kumar-subramanian) [inline doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:1049]: What is the communication mechanism between translib/transformer and bgpcfgd? I do not see this explained anywhere in the document. Also these flow diagrams need explanation.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+
+## 2.2 Functional Description
+Provide GNMI and REST support for config get/set, state get and statistics get,  CLI config and show commands for FRR-BGP features used in SONiC.
+
+
+# 3 Design
+## 3.1 Overview
+The extended unified config and management framework for FRR-BGP in SONiC is represented in below diagram.
+
+![FRR-BGP Unified Mgmt Framework](images/FRR-BGP-Unified-mgmt-frmwrk.png)
```
Comment (anand-kumar-subramanian) [inline doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:148]: The state and statistics flow in this diagram does not match with the CLI and REST get flow diagram. In the flow diagrams the response is through bgpcfgd but in this diagram it is shown that the FRR directly talks to management framework. Only one can be right.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+
+2. Provide annotations for required objects so that transformer core and common app will take care of handling them.
+
+3. Provide transformer methods as per the annotations defined to take care of model specific logics and validations.
+
+4. Define SONiC YANG and Redis ABNF schema for the supported Open Config BGP models & objects.
+
+5. KLISH CLI and REST clients provide extensive BGP configurations and hence there should not be any need for BGP configurations via vtysh.
+
+6. In bgpcfgd register for Redis DB events for the BGP and other related objects, so as to translate the Redis DB events to FRR-BGP CLI commands to configure FRR-BGP, similarly, separate config daemons can be present to configure individual features like OSPF, BFD..etc
```
Comment (anand-kumar-subramanian) [inline doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:160]: By FRR-BGP CLI command, we mean vtysh CLI commands right? If so will the vtysh shell be opened once from bgpcfgd or will it be opened for every single configuration command?

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md [hunk 4]
```diff
@@ -0,0 +1,879 @@
...
+
+7. Update /usr/share/sonic/templates/bgpd.conf.j2 template for new FRR-BGP configurations supported in SONiC which will be used by sonic-cfggen to generate /etc/frr/bgpd.conf file.
```
Comment (anand-kumar-subramanian) [inline doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:162]: Should we include the template here?

**Review by anand-kumar-subramanian — CHANGES_REQUESTED** [review]

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md [hunk 5]
```diff
@@ -0,0 +1,879 @@
...
+
+![FRR-BGP-CONFIG-SEQUENCE](images/frr-bgp-config-sequence.jpg)
+
+## 4.2 CLI Show Command Sequence
+
+### 4.2.1 CLI Show Sequence - Config information
+![FRR-BGP-CLI-SHOW-SEQUENCE1](images/frr-bgp-cli-show-sequence11.jpg)
+
+### 4.2.2 CLI Show Sequence - State/Statistics
+![FRR-BGP-CLI-SHOW-SEQUENCE2](images/frr-bgp-cli-show-sequence22.jpg)
```
Comment (sachinholla) [inline doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:1041]: Diagram shows translib sends a synchronous message to bgpcfgd which in turn runs vtysh command. What is the messaging used between translib and bgpcfgd?

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md [hunk 6]
```diff
@@ -0,0 +1,879 @@
...
+
+  3. The Swagger generated REST server handles all the REST requests from the client SDK and invokes a common handler for all the create, update, replace, delete and get operations along with path and payload. This common handler converts all the requests into Transformer arguments and invokes the corresponding Transformer APIs.
+
+  4. For show commands, the CLI request is converted to a corresponding REST client SDK get request based on Open Config data model's config or state object on a case by case basis.
+
+  5. For show commands that requires retrieval of the data that doesn't contain any state information (information only based on the configuration), the backend callback will fetch the data from CONFIG_DB.
+
+  6. For show commands that requires retrieval of state or statistics information the backend callback will fetch the data from FRR-BGP.
+
+  7. State/stats information is retried from FRR-BGP by issuing a show command to FRR BGP container and the output is returned in JSON format.
```
Comment (sachinholla) [inline doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md]: Not in sync with flow diagram 4.2.2 - it shows translib sends a synchronous message to bgpcfgd to fetch state data.
Also, isn't direct container-to-container communication against SONiC design guidelines?

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 1 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:1049]: Updated the doc with "docker exec" command.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 2 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:148]: will fix this.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 3 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:160]: yes, will check the possibility in bgpcfgd and make the changes accordingly.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 4 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:162]: No, we're just enhancing the existing bgpd.conf.j2 for new configs.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 5 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:1041]: will fix it.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 6 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md]: will fix it

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 1 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:1049]: There is a mention of event handling in the diagram section 4.1

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 2 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:148]: Fixed the diagram, bgpcfgd is not involved for state/stats get.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 3 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:160]: will check and fix accordingly.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 4 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:162]: No, this is well known template file and already being used for BGP neighbors, peer range..etc

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 5 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md:1041]: Fixed the diagram, there is no bgpcfgd involvement for state/stats show.

File: doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md (hunk 6 above)
Comment (venkatmahalingam) [reply doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md]: This is fixed now.

**Review by anand-kumar-subramanian — APPROVED** [review]

**Pull request merged**